### ISSUE TITLE
Clock multiplier less than 0.5 feature request

### ISSUE THREAD

**Issue by sicklittlemonkey** [AppleWin/AppleWin#1254]
User [Cosmos Soup](https://www.facebook.com/groups/5251478676/user/100064451725387/) on [Facebook](https://www.facebook.com/groups/5251478676/posts/10163092909413677/):

> AppleWin won't go below 500 kilohertz. I'm looking for something that will at least get down into the 1/10 normal speed range if not slower.

And:

> I have some 6502 source that is hard to follow, and the description of how it works doesn't seemingly fit with what I'm seeing onscreen. However, 60Hz video refresh rates limit our ability to see 1MHz screen-RAM character changes reliably. By slowing it way down, I will be better able to see what's the executable is doing onscreen. That should help me determine better how the code actually works.

I had a quick look at the code, and see multiple checks ensuring the lower limit of 0.5, and relying on that - so it's not trivial to change. I presume those checks are related to keeping the UI consistent. It seems possible though, and the request seems reasonable for development.

**Comment by sicklittlemonkey** [comment]
Just FYI, I did a custom build for "Cosmos Soup" with the following change:
Core.cpp line 219:
```
		g_fMHz = 0.5 + (double)g_dwSpeed * 0.05;
->
		g_fMHz = 0.05 + (double)g_dwSpeed * 0.05;
```

This successfully shifts the lower clock speed range from 0.5-0.95 to 0.05-0.5 MHz.
Going lower than 0.05 triggers an assert in the sound code which I haven't investigated.

**Comment by sicklittlemonkey** [comment]
"Cosmos Soup" gave the following feedback:
> Now I ran into what might be another problem- one for you to sleep on perhaps.  I am seeing the digits that are written to the screen, as the program I'm exploring normally does. SO far so good.  But I am seeing at times, only 2 or 3 virtual scanlines for a character "cell", making it impossible to see what it's doing (outputting to the screen).
>
> That is, as the normal flurry of digits flood the screen, it's unreadable. At slower speed, it may help me actually see what the program is doing, so then hopefuly the assembler source makes more sense. The run speed is indeed much lower. But now the digits aren't being made entirely visible. Odd.

I've asked him if he can share the program, or if not, a program with similar behaviour for testing.

He mentioned that it used text mode, so after having a play around I'm guessing he's seeing some tearing. The rendering logic is a lot more complex than it was 20 years ago, and I'm not very familiar with it any more. Decoupling rendering from slowed execution would change this feature request difficulty to Epic, but I noticed something else that might help.

Sometimes the Scroll-Lock Full Speed feature has a bug (which I haven't solved). When in this buggy state, pressing Scroll-Lock doesn't change the speed - but it does change the rendering mode. The tearing disappears; so some kind of full-screen updates are being done. Since speed isn't an issue, could this be forced at lower speeds? What do you guys think? 

Here's a toy program to play with - careful, it stomps on screen holes, needs reset etc:
```
0300-   BD 00 04    LDA   $0400,X
0303-   49 FF       EOR   #$80
0305-   9D 00 04    STA   $0400,X
0308-   E8          INX   
0309-   4C 00 03    JMP   $0300
```
300:BD 00 04 49 80 9D 00 04 E8 4C 00 03

This one was at normal speed.
![image](https://github.com/AppleWin/AppleWin/assets/4393005/f4064ffc-c67c-4064-a326-c80d2ec51a9c)

In any case, it might be nice if AppleWin supported slower execution. My hack is probably better for the UI as:
```
g_fMHz = 0.05 + (double)g_dwSpeed * 0.1;
Or
g_fMHz = 0.1 + (double)g_dwSpeed * 0.1; // This would look cleaner, but he did specifically ask for slower than 1/10
```

The idea being that in the UI only the minimum speed text needs to be changed. It would be nice if the UI or command line option supported arbitrary values, but that would take more work.

**Comment by tomcw** [comment]
Let me feedback on some of the points...

>Going lower than 0.05 triggers an assert in the sound code which I haven't investigated.

I tried with 0.04. The first assert I hit was in `ReinitRemainderBuffer()`.

`SetClksPerSpkrSample()` calculates a global variable `g_fClksPerSpkrSample`, and going lower than 0.05 (50kHz(*)) means that `g_fClksPerSpkrSample` gets set to 0.0.

(*) NB. A value of 0.05 gives a clock freq of 50kHz, not 500kHz.

re. command line
There is already a command line switch "-clock-multiplier", which is clamped by `UseClockMultiplier()` to be the same as the Config GUI, ie. x0.5 to x3.9. So we could just relax the command line / `UseClockMultiplier()` to allow going as low as 0.05. 

....But then the Config GUI wouldn't accurately reflect this very low clock freq.
So we could change the Config GUI to show: `0.05   0.5   1.0   2.0           Fastest`
Currently it shows: `0.5   1.0   2.0           Fastest`

re. Scroll-Lock Full Speed
From the Config GUI (Input tab), there are 2 modes for Scroll Lock: either only run at full-speed when Scroll-Lock is pressed (unchecked, default) or act as a toggle for full-speed (checked).

See the main-loop - `ContinueExecution()` - when `g_bFullSpeed == true`, then it just does video updates once every 17ms (wall-clock), ie. here: `VideoRedrawScreenDuringFullSpeed()`.

If you can repro the bug, then it'd be good to fix it.

**Comment by tomcw** [comment]
A different approach (depending on what "Cosmos Soup" is actually trying to do), is to use the built-in debugger.

When you are stopped (eg. hit a breakpoint or just enter debugger with F7), then using the `text` command will do a full-screen render, so no tearing.

And also the "Pause" key will pause, and do a full-screen render, so again no tearing visible.

**Comment by sicklittlemonkey** [comment]
Thanks Tom, for your concise and helpful (as always!) comments.

Firstly, the thing I should have asked first: the program being debugged is an oldie but a goodie!
[Sudoku Solver](http://www.michaeljmahon.com/Sudoku.html) by Michael J. Mahon & Scott Hemphill;

Boot, hit any key, Ctrl-L, then load RVEVIL, and hit Return to start solving it.
It quickly becomes apparent why you might want to slow this down.

> g_fClksPerSpkrSample gets set to 0.0.

Excuse my brutal edits. It's running slow, so we don't care how it sounds. ; - )
I commented out the newer line:
`//g_fClksPerSpkrSample = (double) (UINT) (g_fCurrentCLK6502 / (double)SPKR_SAMPLE_RATE);`
And uncommented the older one:
`g_fClksPerSpkrSample = g_fCurrentCLK6502 / (double)SPKR_SAMPLE_RATE;`
And then made sure it had a little buffer:
`g_nRemainderBufferSize = MAX(1, (UINT) g_fClksPerSpkrSample);`

> There is already a command line switch "-clock-multiplier"

Yeah, but even the UI is a bit cumbersome, and I wanted to make it easy for this guy.
So I overloaded F9 and F10 (I know, I know) but now with Ctrl they clock it down/up.
And I added the speed to the Window's title bar, because it was easy and necessary:
![image](https://github.com/AppleWin/AppleWin/assets/4393005/5b4e62e2-d51f-4f31-8e14-6f1e5a5ce063)

As for the pre-determined speeds, I wanted to play with them, so ...
```
	if (g_dwSpeed < SPEED_NORMAL)
	{
		switch (g_dwSpeed)
		{
		case 0: g_fMHz = 0.0001; break;
		case 1: g_fMHz = 0.0005; break;
		case 2: g_fMHz = 0.0010; break;
		case 3: g_fMHz = 0.0050; break;
		case 4: g_fMHz = 0.0100; break;
		case 5: g_fMHz = 0.0500; break;
		case 6: g_fMHz = 0.1000; break;
		case 7: g_fMHz = 0.2000; break;
		case 8: g_fMHz = 0.4000; break;
		case 9: g_fMHz = 0.8000; break;
		}
	}
```

> re. Scroll-Lock Full Speed

Yes, I find both modes very useful, but for this, the toggle is better. (I'll try to repro the bug.)

> See the main-loop - ContinueExecution() - when g_bFullSpeed == true, then it just does video updates once every 17ms (wall-clock), ie. here: VideoRedrawScreenDuringFullSpeed().

Thanks, this confirmed that I had already played with exactly the right lines of code, but somehow missed the redraw difference. I forced full redraw for the slow speeds, but really this would need to be another option. I know we have too many options already, but try looking at the solver running very slowly with "partial" updates. Decoupling the redraw makes sense here:
`if (g_bFullSpeed || (g_dwSpeed < SPEED_NORMAL))`

And at slower speeds it doesn't get through every 17 ms because not enough cycles, so:
```
	UINT dwClksPerFrame = NTSC_GetCyclesPerFrame();
	if (g_dwSpeed < SPEED_NORMAL)
		dwClksPerFrame = (UINT)(dwClksPerFrame * g_fMHz);
```

Thanks for the note about Pause. That is useful, but gets repetitive ... it's nicer to let it flow, but slow enough to see. And I did originally suggest the debugger, but that becomes even more cumbersome. It would be nice if it returned to the display mode (e.g. text) automatically.

My changes are trivial, but I could create a branch if you want me to. A binary is [here](https://drive.google.com/file/d/1YdtkBGH5ivbiqv-UIo9SrEX3zMHFlZZU/view?usp=drive_link).

Cheers,
Nick.

**Comment by tomcw** [comment]
>Excuse my brutal edits. It's running slow, so we don't care how it sounds. ; - )

Right! ...so we could just silence the speaker below a certain speed (eg. maybe at 0.5MHz it's useful to hear it, but anything slower then silence).

>As for the pre-determined speeds, I wanted to play with them [...]

I don't think we need all these sub-1MHz speeds - maybe just pick a few?

And I don't think we need more keyboard shortcuts to switch to different speeds (but maybe this change is just for your experimentation?). BTW, we already have Ctrl-0 (Toggle between custom speed and Full-Speed ), Ctrl-1 (1MHz) and Ctrl-3 (Full-Speed).

Or perhaps changing the Config GUI is better?
Instead of a (non-linear) slider, then a drop-down and/or an edit control (for a custom speed).

**Comment by sicklittlemonkey** [comment]
> so we could just silence the speaker below a certain speed

Although we don't care how it sounds, it's actually handy to hear the clicks. At least then we know what it's trying to do. But I suppose it also depends on whether changes to the sound code are permissible - similar to what I did to get it to work.

> I don't think we need all these sub-1MHz speeds - maybe just pick a few?
> Or perhaps changing the Config GUI is better?
> Instead of a (non-linear) slider, then a drop-down and/or an edit control (for a custom speed).

There were already 10 levels, so I just picked 10 values. It's fun being able to change it without going to the config, but yeah, I didn't know what the user wanted. It's only one user at this stage, and nobody else has chimed in. He did say:
>The Ctrl F9 and F10 very handy, and when using those, the speed updates on the drag bar. The fine control over lower speeds I find very helpful.

And also:
More feedback via FB chat:
>Wow this is amazeballs!  I had it slow enough where I could start to visually follow it for the first time!
>This is the coolest thing ever. Speed overrated sometimes 🙂
>This is amazing, watching a cat happen and text gets spit out at like 300 baud. Just like 1985 being on the early BBSes.

In real world debugging, defining slow, and being able to toggle sounds good enough. A drop down and text field would provide both the convenience of a predefined list and fine control. And that way you could see what the speed is - which we currently couldn't do, hence the title bar hack.

He also said:
>Obviously the option of the video display not being reliant on 6502 speed is a huge benefit when using AppleWin to observe screen output.

This is something that smells like an option, which I know you don't want. ; - ) For this case the desired behaviour was full redraw, but imagine examining a French Touch demo etc. Unlikely perhaps.

Anyway, it's remarkable that it works as well as it does when dialed down dog-slow.
I was surprised I didn't run into more problems, though more testing might. It's very cool.

**Comment by tomcw** [comment]
Let me summarise the work for this request:

- [ ] CUI: allow slower CPU speeds to a minimum of 0.05MHz (=50kHz).
   - continue to allow speaker sound at this very slow speed.
   - NB. anything slower than this, and the speaker either needs to be silenced or the implementation needs a more invasive change.
- [ ] CUI: a new option to do periodic full frame video updates (like we already have for full-speed mode).
- [ ] GUI (Config): CPU speed: change from (non-linear) slider to a drop-down and/or an edit control (for a custom speed).
